test/distribution/test_nan_mixture.py

```diff
@@ -0,0 +1,344 @@
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Standard library imports
+from typing import Iterable, List, Tuple
+
+# Third-party imports
+import mxnet as mx
+import numpy as np
+import pytest
+
+# First-party imports
+from gluonts.gluonts_tqdm import tqdm
+from gluonts.model.common import Tensor, NPArrayLike
+from gluonts.mx.distribution.distribution import Distribution
+
+from gluonts.mx.distribution import (
+    Categorical,
+    CategoricalOutput,
+    Gaussian,
+    GaussianOutput,
+    NanMixture,
+    NanMixtureOutput,
+    StudentTOutput,
+)
+from gluonts.core.serde import dump_json, load_json
+
+serialize_fn_list = [lambda x: x, lambda x: load_json(dump_json(x))]
+
+
+def diff(x: NPArrayLike, y: NPArrayLike) -> np.ndarray:
+    return np.mean(np.abs(x - y))
+
+
+NUM_SAMPLES = 2_000
+NUM_SAMPLES_LARGE = 100_000
+
+p = np.array([[[0.0001, 0.0001, 0.5]], [[0.999, 0.999, 0.5]]])
+p_cat = np.array(
+    [
+        [[[0.1, 0.9], [0.9, 0.1], [0.5, 0.5]]],
+        [[[0.9, 0.1], [0.01, 0.99], [0.45, 0.55]]],
+    ]
+)
+SHAPE = p.shape
+
+
+@pytest.mark.parametrize(
+    "distr, cat_distr, p, p_cat",
+    [
+        (
+            Gaussian(
+                mu=mx.nd.zeros(shape=SHAPE),
+                sigma=1e-3 + mx.nd.ones(shape=SHAPE),
+            ),
+            Categorical(log_probs=mx.nd.log(mx.nd.array(p_cat))),
+            mx.nd.array(p),
+            mx.nd.array(p_cat),
+        ),
+    ],
+)
+@pytest.mark.parametrize("serialize_fn", serialize_fn_list)
+def test_nan_mixture(
+    distr: Distribution,
+    cat_distr: Distribution,
+    p: Tensor,
+    p_cat: Tensor,
+    serialize_fn,
+) -> None:
+    # sample from component distributions, and select samples
+
+    samples = distr.sample(num_samples=NUM_SAMPLES_LARGE)
+
+    rand = mx.nd.random.uniform(shape=(NUM_SAMPLES_LARGE, *p.shape))
+    choice = (rand > p.expand_dims(axis=0)).broadcast_like(samples)
+    samples_ref = mx.nd.where(choice, samples, samples.zeros_like() / 0.0)
+
+    # construct NanMixture distribution and sample from it
+    nan_mixture = NanMixture(nan_prob=p, distribution=distr)
+    nan_mixture = serialize_fn(nan_mixture)
+
+    samples_mix = nan_mixture.sample(num_samples=NUM_SAMPLES_LARGE)
+
+    # check that shapes are right
+
+    assert samples.shape == samples_mix.shape == samples_ref.shape
+
+    # TODO check mean and stddev
+
+    x = mx.nd.array([[[np.nan, 3.5, -0.5]], [[np.nan, 3.5, np.nan]]])
+
+    # check log_prob
+    log_prob = nan_mixture.log_prob(x)
+    log_prob_true = mx.nd.log(mx.nd.where(x != x, p, (1 - p) * distr.prob(x)))
+
+    assert np.allclose(log_prob.asnumpy(), log_prob_true.asnumpy(), atol=1e-1)
```

Comment:
I’m wondering if these allclose could use lower tolerance (or just the default one). Same for other calls below in this same test. Are expecting these two values to be “somewhat close” or essentially the same?

Reply:
Yes, thanks. Default 1e-8 makes sense and works.